File: android/moqkit/moqkit/src/main/java/com/swmansion/moqkit/publish/source/FrameSource.kt
package com.swmansion.moqkit.publish.source

import android.view.Surface

/**
 * Advanced extension point for custom video capture sources.
 *
 * Most apps should use [CameraCapture], [MultiCameraCapture], or [ScreenCapture]. Implement
 * this interface only when the app needs to feed frames from its own rendering or capture
 * pipeline.
 */
interface VideoFrameSource {
    /**
     * Connects the source to the encoder input surface.
     *
     * [com.swmansion.moqkit.publish.Publisher] calls this when a video track starts.
     * Sources that assign presentation timestamps explicitly must use the monotonic
     * [System.nanoTime] timebase so audio and video remain comparable. Leaving timestamps
     * automatic on an Android [Surface] already uses that timebase.
     */
    fun attachEncoderSurface(surface: Surface)

    /**
     * Disconnects the encoder surface.
     *
     * [com.swmansion.moqkit.publish.Publisher] calls this when a video track stops.
     */
    fun detachEncoderSurface()

    /**
     * Sets an optional preview surface for local display.
     *
     * Pass `null` to clear the preview.
     */
    fun setPreviewSurface(surface: Surface?)
}

/**
 * Advanced extension point for custom audio capture sources.
 *
 * Most apps should use [MicrophoneCapture]. Implement this interface only when the app
 * needs to publish PCM samples from another source.
 */
interface AudioFrameSource {
    /**
     * Callback that receives PCM 16-bit audio samples.
     *
     * [com.swmansion.moqkit.publish.Publisher] sets this when an audio track starts and
     * clears it when the track stops. The timestamp is the first sample's capture time in
     * microseconds using the monotonic [System.nanoTime] timebase.
     */
    var onPcmData: ((data: ByteArray, size: Int, timestampUs: Long) -> Unit)?
}

File: ios/Sources/MoQKit/Publish/source/FrameSource.swift
import CoreMedia

/// Advanced protocol for custom capture sources that feed `CMSampleBuffer` values into ``Publisher``.
///
/// Most apps can use built-in sources such as ``CameraCapture``, ``MicrophoneCapture``,
/// and ``ScreenCapture``. Conform to `FrameSource` when you already have your own capture
/// pipeline and want to plug it into MoQKit publishing.
public protocol FrameSource: AnyObject, Sendable {
    /// Callback installed by the publisher when the track starts.
    ///
    /// Call this for every captured sample buffer. Return `false` from the callback means
    /// the downstream consumer no longer wants frames, so the source should stop producing them.
    var onFrame: (@Sendable (CMSampleBuffer) -> Bool)? { get set }
}
